be/dtm0_log.c

```diff
@@ -0,0 +1,284 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_DTM
+#include "be/dtm0_log.h"
+#include "be/list.h"
+#include "lib/assert.h" /* M0_PRE */
+#include "lib/errno.h"  /* ENOENT */
+#include "lib/memory.h" /* M0_ALLOC */
+
+enum {
+	dl_magix = 0xdeaddeaddeaddead
+};
+
+static const struct m0_tl_descr dldesc = M0_TL_DESCR("DTM0 Log Record",
+                                                     struct m0_dtm0_log_rec,
+                                                     dlr_tlink,
+                                                     dlr_magic,
+                                                     dl_magix,
+                                                     0xbadbadbadbadbadb);
+
+M0_INTERNAL bool m0_be_dtm0_log__invariant(const struct m0_be_dtm0_log *log)
+{
+	return _0C(log != NULL) &&
+	       _0C(log->dl_cs != NULL) &&
+	       _0C(m0_tlist_invariant(&dldesc, log->dl_tlist));
+}
+
+M0_INTERNAL bool m0_dtm0_log_rec__invariant(const struct m0_dtm0_log_rec *rec)
+{
+	return _0C(rec != NULL) &&
+	       _0C(m0_dtm0_tx_desc__invariant(&rec->dlr_txd)) &&
+	       _0C(m0_tlink_invariant(&dldesc, rec));
+}
+
+M0_INTERNAL void m0_be_dtm0_log_init(struct m0_be_dtm0_log *log,
+                                     struct m0_dtm0_clk_src *cs)
+{
+	M0_PRE(cs);
+	M0_PRE(log);
+	m0_mutex_init(&log->dl_lock);
+	m0_tlist_init(&dldesc, log->dl_tlist);
+	log->dl_cs = cs;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_fini(struct m0_be_dtm0_log *log)
+{
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	m0_mutex_fini(&log->dl_lock);
+	m0_tlist_fini(&dldesc, log->dl_tlist);
+	log->dl_cs = NULL;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_credit(enum m0_be_dtm0_log_credit_op op,
+                                       struct m0_be_tx              *tx,
+                                       struct m0_be_seg             *seg,
+                                       struct m0_be_tx_credit       *accum)
+{
+	/*TODO:Complete implementation during persistent list implementation */
+	switch (op) {
+	case M0_DTML_CREATE:
+	case M0_DTML_SENT:
+	case M0_DTML_EXECUTED:
+	case M0_DTML_PERSISTENT:
+	case M0_DTML_REDO:
+	default:
+		M0_IMPOSSIBLE("");
+	}
+}
+
+M0_INTERNAL void m0_be_dtm0_log_create(struct m0_be_tx        *tx,
+                                       struct m0_be_seg       *seg,
+                                       struct m0_be_dtm0_log **out)
+{
+	struct m0_be_dtm0_log *log = NULL;
+
+#if 0
+	M0_PRE(tx);
+	M0_PRE(seg);
+	M0_PRE(m0_be_tx__invariant(tx));
+
+	M0_BE_ALLOC_PTR_SYNC(log, seg, tx);
+	M0_ASSERT(log);
+	M0_BE_ALLOC_PTR_SYNC(log->dl_list, seg, tx);
+	M0_ASSERT(log->dl_list);
+
+	dtm0_log_be_list_create(log->dl_list, tx);
+#endif
+	M0_PRE(out);
+	M0_ALLOC_PTR(log);
+	M0_ASSERT(log);
```

Comment:
Let's return -ENOMEM (`return M0_ERR(-ENOMEM)`) here and in the other places.

Reply:
Done.